Enable Up/Down reference for artificial horizon in OSD

bobbycisneros opened the pull request:
This mod adds an arrow symbol showing the location of the up/down on the OSD.  The arrow direction points in the direction of up or down.

See test video:

https://www.youtube.com/watch?v=bo84m2KZtq8&list=PLBdJNyPDcSIeXDO8PsFbzhbAsvIaRKmm0&index=17

etracer65 commented:
Looking at the video I can't really determine what the "arrow" is trying to show. It just seems to jump around randomly. Most of the time the quad appears to be pointing straight up so not sure what the arrow is trying to indicate and how it relates to "up/down".

Beyond that, not sure why we need such a thing. Why would a typical Betaflight user flying a FPV quad need an "up/down" reference? After all, we have a camera that's showing the orientation to the pilot. It's not like we should spend much time in an orientation that would be ambiguous (like pointing straight up) as the quad would be falling from the sky. And I don't see how somebody would have the reactions to look for some indicating arrow and use that to recover in a situation like this in the timeframe required.

Next you've enabled this extra "arrow" for everybody who wishes to use the artificial horizon OSD element. Based on the points above it seems unlikely that many people would actually want this so forcing it on for everybody is not the proper way to go.

Lastly, this is an awful lot of code (and resulting flash usage) for such a simple thing. Before this could be considered (assuming better justification of value) the code would need to be significantly optimized.

bobbycisneros commented:
Hey Bruce,

Thanks for the constructive feedback!  

The purpose of my update is to provide a visual reference to vertical.  When flying, often times I find myself pointing up to sky but no other reference to how far over I am.  Having this visual reference helps with more precise power loops.  Also, during  pushback maneuvers where a high pitch attitude is required its handy to know if the drone is <90 degrees to ensure I am pushing back with the drone thrust vector slightly down to counter gravity.  The visual reference to down was a simple addition.  

The location of the arrow provides a visual reference as to the location of vertical (or down) w/re to the local level frame.  The arrow points to the direction from the crosshairs (bore site) to vertical. The motion appears to step due to the slow OSD update rate and the fact the OSD only shows characters (not a calligraphic or raster display like that of aircraft heads up displays); however it is not random. 

Yes, I enabled it for everybody who uses the artificial horizon because I thought it made sense to be part of that suite of symbology.  If this should be its own feature then I can make it so but; being a newbie, I have no idea how to add this as a separate feature that can be activated via the configurator.  I downloaded the configurator project but I am not familiar with Java nor can I get the project to compile so I’m stuck there.  The instructions are way too lacking on how to compile that project so I need some help there.  I may figure out how to add a CLI parameter as a first step if you think it needs to be a separate feature.  Can you point me to a resource so I can figure out how to compile the configurator project?  

As to the code size, I can review it to see how it can be streamlined.  Most of the code is for figuring out which way to point the arrow and the rest is fairly simple.  I may even be able to eliminate the arcsine calls b/c the sine of a small angle is approximately equal to the angle.  I’ll review/optimize the code.  

Question, is there a way to test changes to the firmware via simulation in lieu of compiling the code and loading it into a drone for test?   In my industry, we use high fidelity 6-dof simulations to test candidate software changes before release to the field.  I’ve seen several good FPV simulators on STEAM.  Is there a way to link in the code with those sims?

Robert
  

> On Dec 22, 2020, at 8:42 AM, Bruce Luckcuck <notifications@github.com> wrote:
> 
> 
> Looking at the video I can't really determine what the "arrow" is trying to show. It just seems to jump around randomly. Most of the time the quad appears to be pointing straight up so not sure what the arrow is trying to indicate and how it relates to "up/down".
> 
> Beyond that, not sure why we need such a thing. Why would a typical Betaflight user flying a FPV quad need an "up/down" reference? After all, we have a camera that's showing the orientation to the pilot. It's not like we should spend much time in an orientation that would be ambiguous (like pointing straight up) as the quad would be falling from the sky. And I don't see how somebody would have the reactions to look for some indicating arrow and use that to recover in a situation like this in the timeframe required.
> 
> Next you've enabled this extra "arrow" for everybody who wishes to use the artificial horizon OSD element. Based on the points above it seems unlikely that many people would actually want this so forcing it on for everybody is not the proper way to go.
> 
> Lastly, this is an awful lot of code (and resulting flash usage) for such a simple thing. Before this could be considered (assuming better justification of value) the code would need to be significantly optimized.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-749573301>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3ALDG2NB5T5XTILLN3SWCV3TANCNFSM4VDIKTUA>.
>

etracer65 commented:
Sorry, I still don't see the value. I think you need to step outside your own implementation and think about how this presentation adds value to the average user. I'm still not sure if the arrow is trying to tell me which way is up, down, or which way I need to move the controls. And for what purpose? To get back to level? Any kind of presentation like this begins to have issues around the 90 degree inflection points. Having an arrow that jumps back and forth in these cases doesn't really add value. There might be some value if first we can:

1. identify a clear problem this is intended to solve
2. show that it's really a problem we need to solve and will be useful to many users
3. make the presentation understandable and generally usable

Since there have been no feature requests for something like this and nobody has tried to implement this before it would seem that item 2 is going to be difficult to justify.

Betaflight has a SITL target that you could integrate with but it has no OSD support so it wouldn't be useful in this case.

bobbycisneros commented:
Ok. No worries. It works for me and I will keep it in my repo in case anybody else is interested. The intention was clearly stated in my previous email and it’s certainly valuable to me but if there is no pull then I understand. 

BTW, I was careful to ensure there would be no problems around the inflection points by transforming the vertical vector to the body frame of the drone. I’m a aerospace engineer and spent years working missile simulations so I understand the math and how to avoid “gimbal” lock. 

Sent from my iPhone

> On Dec 22, 2020, at 1:50 PM, Bruce Luckcuck <notifications@github.com> wrote:
> 
> ﻿
> Sorry, I still don't see the value. I think you need to step outside your own implementation and think about how this presentation adds value to the average user. I'm still not sure if the arrow is trying to tell me which way is up, down, or which way I need to move the controls. And for what purpose? To get back to level? Any kind of presentation like this begins to have issues around the 90 degree inflection points. Having an arrow that jumps back and forth in these cases doesn't really add value. There might be some value if first we can:
> 
> identify a clear problem this is intended to solve
> show that it's really a problem we need to solve and will be useful to many users
> make the presentation understandable and generally usable
> Since there have been no feature requests for something like this and nobody has tried to implement this before it would seem that item 2 is going to be difficult to justify.
> 
> Betaflight has a SITL target that you could integrate with but it has no OSD support so it wouldn't be useful in this case.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ledvinap commented:
Maybe just displaying symbol for/on zenit and nadir will be useful. It will integrate quite cleanly with existing horizon and will be mostly unobtrusive.

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -177,6 +178,8 @@ static unsigned activeOsdElementCount = 0;
 static uint8_t activeOsdElementArray[OSD_ITEM_COUNT];
 static bool backgroundLayerSupported = false;
 
+extern float rMat[3][3];
```
Never declare `extern` in .c file, include `imu.h` instead.
This could lead to very difficult bugs in variable type is changed.

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -599,7 +603,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+   int psiBSign; // psi from body frame to up vector
+   char arrowDir; // direction to draw arrow
+   float arrowDirFloat;
+   if(attitude.values.pitch>0.0){ //nose down
```
Testing sign of `z` of up vector  may be cleaner.
Also, isn't it possible to select displayed vector (up/down vector) and unify other calculations?

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -599,7 +603,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+   char arrowDir; // direction to draw arrow
+   float arrowDirFloat;
+   if(attitude.values.pitch>0.0){ //nose down
+       thetaB = asinf(earthUpinBodyFrame[2]) * (1800.0f / M_PIf); // get theta
```
Using simple projection on (camera) plane may be enough.  Displaying zenit/nadir when it's far from camera direction isn't that useful (and there are other visual clues  on OSD in that case).

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -599,7 +603,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+   }
+   
+   //Select which arrow points to the up vector when pitchinng up or down vector when pitching down
+   if     (arrowDirFloat >= -180.00 && arrowDirFloat < -168.75)arrowDir = SYM_ARROW_SOUTH;
```
Use table (or better equation + arrow table) if this is really necessary

ledvinap commented:
It may be also possible to add up direction to horizon element

bobbycisneros commented:
Zenit is up? Nadir is down?


Sent from my iPhone

> On Dec 23, 2020, at 7:08 AM, Petr Ledvina <notifications@github.com> wrote:
> 
> ﻿
> Maybe just displaying symbol for/on zenit and nadir will be useful. It will integrate quite cleanly with existing horizon and will be mostly unobtrusive.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ledvinap commented:
> Zenit is up? Nadir is down?
Sorry, I'm not native speaker. It sounded to me as best description.

https://en.wikipedia.org/wiki/Zenith

bobbycisneros commented:
Nice. I learned something new. I know zenith but never heard of Nadir. 

Sent from my iPhone

> On Dec 23, 2020, at 2:01 PM, Petr Ledvina <notifications@github.com> wrote:
> 
> ﻿
> Zenit is up? Nadir is down?
> Sorry, I'm not native speaker. It sounded to me as best description.
> 
> https://en.wikipedia.org/wiki/Zenith
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobbycisneros commented:
I simplified the code and replaced the arrow with a simple ‘UP’/‘DN’ when abs(pitch)>60 deg.  The symbology is not as obtrusive any more and the intent is more clearly apparent.  I committed the changes to my repo.  I can create another pull request if you think its useful to other people.  

> On Dec 23, 2020, at 7:36 AM, Petr Ledvina <notifications@github.com> wrote:
> 
> 
> @ledvinap commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r547954456>:
> 
> > @@ -177,6 +178,8 @@ static unsigned activeOsdElementCount = 0;
>  static uint8_t activeOsdElementArray[OSD_ITEM_COUNT];
>  static bool backgroundLayerSupported = false;
>  
> +extern float rMat[3][3];
> Never declare extern in .c file, include imu.h instead.
> This could lead to very difficult bugs in variable type is changed.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r547955890>:
> 
> > @@ -599,7 +603,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>          }
>      }
>  
> -    element->drawElement = false;  // element already drawn
> +   float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +   int thetaB; // pitch from body frame to up vector
> +   int psiBSign; // psi from body frame to up vector
> +   char arrowDir; // direction to draw arrow
> +   float arrowDirFloat;
> +   if(attitude.values.pitch>0.0){ //nose down
> Testing sign of z of up vector may be cleaner.
> Also, isn't it possible to select displayed vector (up/down vector) and unify other calculations?
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r547958370>:
> 
> > @@ -599,7 +603,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>          }
>      }
>  
> -    element->drawElement = false;  // element already drawn
> +   float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +   int thetaB; // pitch from body frame to up vector
> +   int psiBSign; // psi from body frame to up vector
> +   char arrowDir; // direction to draw arrow
> +   float arrowDirFloat;
> +   if(attitude.values.pitch>0.0){ //nose down
> +       thetaB = asinf(earthUpinBodyFrame[2]) * (1800.0f / M_PIf); // get theta
> Using simple projection on (camera) plane may be enough. Displaying zenit/nadir when it's far from camera direction isn't that useful (and there are other visual clues on OSD in that case).
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r547959210>:
> 
> > +   int psiBSign; // psi from body frame to up vector
> +   char arrowDir; // direction to draw arrow
> +   float arrowDirFloat;
> +   if(attitude.values.pitch>0.0){ //nose down
> +       thetaB = asinf(earthUpinBodyFrame[2]) * (1800.0f / M_PIf); // get theta
> +       psiBSign = -1; // correct the sign
> +       arrowDirFloat = atan2_approx(-earthUpinBodyFrame[1], earthUpinBodyFrame[2])*180.0/M_PIf;
> +   }
> +   else{// nose up
> +       thetaB = -asinf(earthUpinBodyFrame[2]) * (1800.0f / M_PIf); // get theta 
> +       psiBSign = 1; // correct the sign
> +       arrowDirFloat = -atan2_approx(-earthUpinBodyFrame[1], -earthUpinBodyFrame[2])*180.0/M_PIf;
> +   }
> +   
> +   //Select which arrow points to the up vector when pitchinng up or down vector when pitching down
> +   if     (arrowDirFloat >= -180.00 && arrowDirFloat < -168.75)arrowDir = SYM_ARROW_SOUTH;
> Use table (or better equation + arrow table) if this is really necessary
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#pullrequestreview-557862246>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3HLWZ7SWC47TRIKCMLSWHW4FANCNFSM4VDIKTUA>.
>

mikeller commented:
@bobbycisneros: Do you have a video for the implemantation of zenith / nadir indication? More problems that I can see when using the artificial horizon element to enable / disable this is that you are not respecting the bounding box of this element, and most users won't be having the artificial horizon centered on screen, so your indication will be offset and in the case of arrows look weird.

bobbycisneros commented:
Michael,

You can find the latest symbology at the link below:

https://www.youtube.com/watch?v=XQmHvOnBcao <https://www.youtube.com/watch?v=XQmHvOnBcao>

On my you tube channel you can see the previous version under the FPV playlist.  

I hadn’t thought of the use case where the attitude display may be offset from the center.  I usually line everything up such that the crosshairs along with the horizon are all level.  Then the crosshairs are on the horizon indicator the drone is level.  That is the way its done on aircraft so I thought this was the most logical configuration.  

> On Dec 24, 2020, at 11:35 AM, Michael Keller <notifications@github.com> wrote:
> 
> 
> @bobbycisneros <https://github.com/bobbycisneros>: Do you have a video for the implemantation of zenith / nadir indication? More problems that I can see when using the artificial horizon element to enable / disable this is that you are not respecting the bounding box of this element, and most users won't be having the artificial horizon centered on screen, so your indication will be offset and in the case of arrows look weird.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-750932972>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3CGUTWM3FKR2CUVGHTSWN3WFANCNFSM4VDIKTUA>.
>

mikeller commented:
Ok, I think the UP/DN version is easier to understand.
Now we need to look what we can do to make this work in a useful way even if artificial horizon and / or crosshairs are not centered - I think the easiest will be to make this into its own element, and allow the user to position it where they want. This will also allow users to compensate for camera tilt.

bobbycisneros commented:
Sounds like a plan. I’ll pick it up again in the coming days. Building a new drone that I got for Christmas and repairing another one. 

Sent from my iPhone

> On Dec 25, 2020, at 6:31 PM, Michael Keller <notifications@github.com> wrote:
> 
> ﻿
> Ok, I think the UP/DN version is easier to understand.
> Now we need to look what we can do to make this work in a useful way even if artificial horizon and / or crosshairs are not centered - I think the easiest will be to make this into its own element, and allow the user to position it where they want. This will also allow users to compensate for camera tilt.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ledvinap commented:
@mikeller: IMO up/down reference should be crosshair symbol (which is probably true for artificial horizon too).

Also ... IMO using single symbol instead of 'UP'/'DN'  will be better choice. Maybe `O` / `X` or some graphical symbol?

bobbycisneros commented:
It seems the consensus is the feature may be useful for some people. I’m good with either up/dn or O/X. X for up? Or the other way, O for up?  

Sent from my iPhone

> On Dec 28, 2020, at 9:19 AM, Petr Ledvina <notifications@github.com> wrote:
> 
> ﻿
> @mikeller: IMO up/down reference should be crosshair symbol (which is probably true for artificial horizon too).
> 
> Also ... IMO using single symbol instead of 'UP'/'DN' will be better choice. Maybe O / X or some graphical symbol?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bobbycisneros commented:
Team,

Updated this new feature but a I still have lingering questions that need to be answered before I do another pull request.  Here are the updates and questions.

1. The feature is now a separate element.  Its position is centered on the display for consistent operation and cannot be changed by the user for consistent operation.  I put a simple “+” for Zenith and “-“ for Nadir.  Its settable via CLI or OSD menu.  
	Question, do all displays have the same number of rows and columns?  I seem to remember that PAL and NTSC were different.  
	If there is a difference, then what are the display sizes?

2. The OSD configurator page now shows the new feature but it is labeled as unknown.
	Question, how do I update the configurator with the right text and tool tip?  
	Is this done through betaflight? 
	Or, is a mod required in the configurator?

You can examine my updates here:

https://github.com/bobbycisneros/betaflight.git <https://github.com/bobbycisneros/betaflight.git>

Regards,
Robert Cisneros


> On Dec 25, 2020, at 6:31 PM, Michael Keller <notifications@github.com> wrote:
> 
> 
> Ok, I think the UP/DN version is easier to understand.
> Now we need to look what we can do to make this work in a useful way even if artificial horizon and / or crosshairs are not centered - I think the easiest will be to make this into its own element, and allow the user to position it where they want. This will also allow users to compensate for camera tilt.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-751303552>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3AUMMMN2DYMHQIHHELSWUVG5ANCNFSM4VDIKTUA>.
>

etracer65 commented:
> Its position is centered on the display for consistent operation and cannot be changed by the user for consistent operation. 

This is not how OSD elements work. Every element can be repositioned by the user. And further this means that your subsequent questions about the row/column sizes are irrelevant as there is no need to "auto" position the element.

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
+// The positions are absolute and immutable by the user for consistent operation
+    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
+        float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
```
`const float ...`

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+        char buf[2]; // character buffer
+
+        if(attitude.values.pitch>0.0){ //nose down
+            thetaB = -asinf(earthUpinBodyFrame[2]) * (180.0f / M_PIf); // get theta
```
keep value in radians (it is scaled later anyway)

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+
+        float thetaB; // pitch from body frame to up vector
+        int psiBSign; // psi from body frame to up vector
+        char buf[2]; // character buffer
```
`char *symbol[2] = {"U","D"};`
It does consume more memory, but is easier to read/understand

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+
+        if(attitude.values.pitch>0.0){ //nose down
+            thetaB = -asinf(earthUpinBodyFrame[2]) * (180.0f / M_PIf); // get theta
+            psiBSign = -1; // correct the sign
```
compute `psiB` here (BTW: using `float` for  `psiBSign` will probably result in faster code)

ledvinap commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+//      Mid    = 14      0
+//      Right  = 28     59
+
+        int posX = round(0.237288136*psiB+14.0); //col
```
`scaleRangef(psiB, -M_PIf / 4, M_PIf / 4, 0, 28)` (using element size instead of 0, 28 and with correct angles) is easier to understand, but a bit slower. Not sure about rounding, using correct bounds may be enough.

Also with custom size, you should check for overflow (IMO one angle shall be limited by attitude.values.pitch, but second is pretty unlimited? -90..90 deg)

bobbycisneros commented:
Thanks Petri, good feedback and I’ll take it into consideration.  

BTW, does BF control the OSD page of the configurator?  I’m trying to figure out how that works and how to get the configurator to properly reflect the option for the new symbology.  Any help you can provide on this would be useful.  

> On Dec 31, 2020, at 12:15 PM, Petr Ledvina <notifications@github.com> wrote:
> 
> 
> @ledvinap commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r550605817>:
> 
> > @@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>  
>      element->drawElement = false;  // element already drawn
>  }
> +
> +static void osdElementUpDownReference(osdElementParms_t *element){
> +// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
> +// The positions are absolute and immutable by the user for consistent operation
> +    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
> +        float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> const float ...
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r550607281>:
> 
> > @@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>  
>      element->drawElement = false;  // element already drawn
>  }
> +
> +static void osdElementUpDownReference(osdElementParms_t *element){
> +// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
> +// The positions are absolute and immutable by the user for consistent operation
> +    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
> +        float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +
> +        float thetaB; // pitch from body frame to up vector
> +        int psiBSign; // psi from body frame to up vector
> +        char buf[2]; // character buffer
> +
> +        if(attitude.values.pitch>0.0){ //nose down
> +            thetaB = -asinf(earthUpinBodyFrame[2]) * (180.0f / M_PIf); // get theta
> keep value in radians (it is scaled later anyway)
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r550609106>:
> 
> > @@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>  
>      element->drawElement = false;  // element already drawn
>  }
> +
> +static void osdElementUpDownReference(osdElementParms_t *element){
> +// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
> +// The positions are absolute and immutable by the user for consistent operation
> +    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
> +        float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +
> +        float thetaB; // pitch from body frame to up vector
> +        int psiBSign; // psi from body frame to up vector
> +        char buf[2]; // character buffer
> char *symbol[2] = {"U","D"};
> It does consume more memory, but is easier to read/understand
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r550610146>:
> 
> > @@ -601,6 +604,54 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>  
>      element->drawElement = false;  // element already drawn
>  }
> +
> +static void osdElementUpDownReference(osdElementParms_t *element){
> +// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
> +// The positions are absolute and immutable by the user for consistent operation
> +    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
> +        float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +
> +        float thetaB; // pitch from body frame to up vector
> +        int psiBSign; // psi from body frame to up vector
> +        char buf[2]; // character buffer
> +
> +        if(attitude.values.pitch>0.0){ //nose down
> +            thetaB = -asinf(earthUpinBodyFrame[2]) * (180.0f / M_PIf); // get theta
> +            psiBSign = -1; // correct the sign
> compute psiB here (BTW: using float for psiBSign will probably result in faster code)
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r550616605>:
> 
> > +
> +//      Symbology Position Calculation 
> +//
> +//      ////// Y ////////
> +//      OSD Pos  Row   ThetaB 
> +//      Top    =  0    -55
> +//      Mid    =  7      0
> +//      Bottom = 15     55
> +
> +//      ////// X ////////
> +//      OSD Pos Col   PsiB  
> +//      Left   =  0    -59
> +//      Mid    = 14      0
> +//      Right  = 28     59
> +
> +        int posX = round(0.237288136*psiB+14.0); //col
> scaleRangef(psiB, -M_PIf / 4, M_PIf / 4, 0, 28) (using element size instead of 0, 28 and with correct angles) is easier to understand, but a bit slower. Not sure about rounding, using correct bounds may be enough.
> 
> Also with custom size, you should check for overflow (IMO one angle shall be limited by attitude.values.pitch, but second is pretty unlimited? -90..90 deg)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#pullrequestreview-560527534>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3H6IZNXPQNNEFWNT5DSXS5U5ANCNFSM4VDIKTUA>.
>

bobbycisneros commented:
The logic behind this was the user will not normally see the up/down indicator on the screen and the positioning of zenith and nadir are absolute.  I’ll change the logic so we can let the user position the reference location for the zenith which will drive the up location on the screen when the drone is aligned with the zenith.  

> On Dec 31, 2020, at 10:14 AM, Bruce Luckcuck <notifications@github.com> wrote:
> 
> 
> Its position is centered on the display for consistent operation and cannot be changed by the user for consistent operation.
> 
> This is not how OSD elements work. Every element can be repositioned by the user. And further this means that your subsequent questions about the row/column sizes are irrelevant as there is no need to "auto" position the element.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-752996982>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3AA6VFYVBYAJSTKRXDSXSPOXANCNFSM4VDIKTUA>.
>

ledvinap commented:
IMO you can copy logic from artificial horizon, it is almost the same problem

ledvinap commented:
BTW: did you try using plane projection (x/z, y/z) for symbol position? with +-45deg range, the difference will be quite small and it may even map better to camera image (camera uses plane projection too)

bobbycisneros commented:
Petri,

For the projection, I could basically dump the arc sine function(s) and use the y and z values of the Zenith directly (scaled to pi/2).  I’ll try it out.  

> On Dec 31, 2020, at 1:02 PM, Petr Ledvina <notifications@github.com> wrote:
> 
> 
> BTW: did you try using plane projection (x/z, y/z) for symbol position? with +-45deg range, the difference will be quite small and it may even map better to camera image (camera uses plane projection too)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-753027910>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3EYWSOE7OHPVXBMXILSXTDC3ANCNFSM4VDIKTUA>.
>

mikeller commented:
@bobbycisneros: Re configurator, the new element will have to be added to configurator manually, but it's relatively simple, see https://github.com/betaflight/betaflight-configurator/pull/1826 for an example.

bobbycisneros commented:
OK team.  Please see my latest commit.  As far as I know addressed all the issues and recommendations brought up to date.  If I am missing anything please advise.  Let me know if there is any other action required to get this approved.  And, thanks for all the help!

bobbycisneros commented:
Please see my latest commit. As far as I know addressed all the issues and recommendations brought up to date. If I am missing anything please advise. Let me know if there is any other action required to get this approved. And, thanks for all the help!  I also made a complimentary change to the betaflight-configurator and created a pull request for that project.  

> On Dec 31, 2020, at 5:09 PM, Michael Keller <notifications@github.com> wrote:
> 
> 
> @bobbycisneros <https://github.com/bobbycisneros>: Re configurator, the new element will have to be added to configurator manually, but it's relatively simple, see betaflight/betaflight-configurator#1826 <https://github.com/betaflight/betaflight-configurator/pull/1826> for an example.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-753222778>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3DP63OBM4HTN7F3BETSXUAEFANCNFSM4VDIKTUA>.
>

mikeller commented on src/main/flight/imu.c:
```diff
@@ -99,7 +99,7 @@ static float smallAngleCosZ = 0;
 static imuRuntimeConfig_t imuRuntimeConfig;
 
-STATIC_UNIT_TESTED float rMat[3][3];
+ float rMat[3][3];
```
Extra whitespace. :space_invader:

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -177,6 +178,8 @@ static unsigned activeOsdElementCount = 0;
 static uint8_t activeOsdElementArray[OSD_ITEM_COUNT];
 static bool backgroundLayerSupported = false;
 
+//extern float rMat[3][3];
```
This should be removed.

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -577,8 +582,8 @@ static void osdElementAntiGravity(osdElementParms_t *element)
-static void osdElementArtificialHorizon(osdElementParms_t *element)
-{
+
+static void osdElementArtificialHorizon(osdElementParms_t *element){
```
Coding style in Betaflight is 1TBS - opening brackets go onto the next line for function definitions.

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +606,37 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+
+// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
+
+    if(abs(attitude.values.pitch)>450){ //pitch is greater than +45 or less than -45
```
Missing whitespace. :space_invader:

mikeller replied:
Using defines instead of magic numbers will remove the need for comments.

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +606,37 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+            thetaB = -earthUpinBodyFrame[2]; // get pitch w/re to nadir (use small angle approx for sine)
+            psiB = -earthUpinBodyFrame[1]; // calculate the yaw w/re to nadir (use small angle approx for sine)
+            direction = DOWN;
+        }
```
Extra newline.

mikeller commented on src/main/osd/osd_elements.c:
```diff
@@ -601,6 +606,37 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
+            direction = UP;
+        }
+
+        int posX = element->elemPosX + round(scaleRangef(psiB, -M_PIf / 4, M_PIf / 4, -14, 14));
```
This element is spanning almost the entire screen size, making it hard for users to position it, and interfering with most other elements - the effecive area should be limited to make it less obtrusive and easier to handle.

bobbycisneros replied:
Mike, thanks for the feedback.  The symbology is normally out of view until high pitch angles  On this issue, the symbology is only in view when pitch exceeds +/- 45.  I only see it during power loops, snap backs etc.  During normal flight when the pilot would have the time to check other elements the symbology is out of view.  If you think it makes sense, I can bound it from the edges.  Please advise if you think that is needed.

mikeller replied:
Yes, I think it makes sense to add a bounding box to this, just like there is a bounding box for every other element. Things like being able to position this on the screen (required to compensate for camera angle) make it necessary to be able to to position this.

bobbycisneros commented:
Michael,

I updated the element logic so it appears when about 25 degrees off Zenith/Nadir effectively bounding it to this area and centered on the location the user selects in the configurator.  Hoping this works.  

Robert

> On Jan 15, 2021, at 8:07 PM, Michael Keller <notifications@github.com> wrote:
> 
> 
> @mikeller commented on this pull request.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r558761750>:
> 
> > +        float psiB; // psi from body frame to zenith/nadir
> +        char *symbol[2] = {"U", "D"}; // character buffer
> +        int direction;
> +
> +        if(attitude.values.pitch>0.0){ //nose down
> +            thetaB = -earthUpinBodyFrame[2]; // get pitch w/re to nadir (use small angle approx for sine)
> +            psiB = -earthUpinBodyFrame[1]; // calculate the yaw w/re to nadir (use small angle approx for sine)
> +            direction = DOWN;
> +        }
> +        else{ // nose up
> +            thetaB = earthUpinBodyFrame[2]; // get pitch w/re to zenith (use small angle approx for sine)
> +            psiB = earthUpinBodyFrame[1]; // calculate the yaw w/re to zenith (use small angle approx for sine)
> +            direction = UP;
> +        }
> +
> +        int posX = element->elemPosX + round(scaleRangef(psiB, -M_PIf / 4, M_PIf / 4, -14, 14));
> Yes, I think it makes sense to add a bounding box to this, just like there is a bounding box for every other element. Things like being able to position this on the screen (required to compensate for camera angle) make it necessary to be able to to position this.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#discussion_r558761750>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3FCEBEFOP4IMISUMRLS2DYFRANCNFSM4VDIKTUA>.
>

mikeller reviewed (commented):
Please rebase and squash into one commit.

bobbycisneros commented:
Mike,

I fixed the coding style and rebased (I think, git is new to me, I have used svn in the past).  Let me know if it all looks ok.  I appreciate your patience with me being a newbie.  Hope to contribute more in the future. 

Noticed a failure in one of the ’test-all’ build.  Not sure why we are suddenly getting an error in that build.  It has worked multiple times in the past.  

BTW - When will there be a build target for the FOXEERF722V2?  

Robert

> On Jan 16, 2021, at 4:08 AM, Michael Keller <notifications@github.com> wrote:
> 
> 
> @mikeller commented on this pull request.
> 
> Please rebase and squash into one commit.
> 
> In src/main/osd/osd_elements.c <https://github.com/betaflight/betaflight/pull/10417#discussion_r558834316>:
> 
> > @@ -601,6 +607,35 @@ static void osdElementArtificialHorizon(osdElementParms_t *element)
>  
>      element->drawElement = false;  // element already drawn
>  }
> +
> +static void osdElementUpDownReference(osdElementParms_t *element)
> +{
> +// Up/Down reference feature displays reference points on the OSD at Zenith and Nadir
> +    const float earthUpinBodyFrame[3] = {-rMat[2][0], -rMat[2][1], -rMat[2][2]}; //transforum the up vector to the body frame
> +
> +    if(ABS(earthUpinBodyFrame[2]) < SINE_25_DEG && ABS(earthUpinBodyFrame[1]) < SINE_25_DEG){ 
> Betaflight coding style is to have spaces between if and ( and between ) and {.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#pullrequestreview-569850564>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3EAXA22QI4OAWY6LJTS2FQQNANCNFSM4VDIKTUA>.
>

mikeller commented:
@bobbycisneros: Uh oh, looks like the rebase has not worked as intended - there are now 25 commits showing instead of the desired 1, and it looks like you have merged some recent change from `master` into your pull request.
I'd offer to help clean this up, but unfortunately you used the `master` branch of your fork for the pull request (instead of creating a dedicated branch for it as is recommended in git) - this means that I cannot push the result of a cleanup back into the branch as a maintainer.

I suspect what you have done is merge the tip of `master` into your branch, which is not a good approach in git.

Instead, what you have to do is:
1. make sure your checked out version of the betaflight `master` branch is up to date;
2. go into your pull request branch;
3. do a `git rebase -i <betaflight master branch>`. You will get a page with a list of commits in your branch. Go through the list and:
- leave the first of your own commits as it is;
- for every other of **your own** commits for this pull request, replace the `pick` with a `squash`;
- for every commit that isn't part of the pull request (hence coming from your attempted rebase): replace the `pick` with `drop`;
4. save the page - the rebase will start;
5. if there are errors due to conflicts, git will stop with an error message;
6. edit the affected files to resolve the conflicts, then do a `git add .; git rebase --continue` and go back to 5.;
7. once the rebase has completed, use `git push --force-with-lease` **important** to force push the changes into your branch

mikeller commented:
@bobbycisneros: Sorry, overlooked your other question. There already is a target for the FOXEERF722V2 - it is the STM32F7X2 Unified Target. Instructions on how to install a custom build of it with the correct hardware configuration are here: https://youtu.be/I1uN9CN30gw

bobbycisneros commented:
Mike,

I think I got it done.  Thanks again for your patience.

Robert

> On Jan 16, 2021, at 3:54 PM, Michael Keller <notifications@github.com> wrote:
> 
> 
> @bobbycisneros <https://github.com/bobbycisneros>: Uh oh, looks like the rebase has not worked as intended - there are now 25 commits showing instead of the desired 1, and it looks like you have merged some recent change from master into your pull request.
> I'd offer to help clean this up, but unfortunately you used the master branch of your fork for the pull request (instead of creating a dedicated branch for it as is recommended in git) - this means that I cannot push the result of a cleanup back into the branch as a maintainer.
> 
> I suspect what you have done is merge the tip of master into your branch, which is not a good approach in git.
> 
> Instead, what you have to do is:
> 
> make sure your checked out version of the betaflight master branch is up to date;
> go into your pull request branch;
> do a git rebase -i <betaflight master branch>. You will get a page with a list of commits in your branch. Go through the list and:
> leave the first of your own commits as it is;
> for every other of your own commits for this pull request, replace the pick with a squash;
> for every commit that isn't part of the pull request (hence coming from your attempted rebase): replace the pick with drop;
> save the page - the rebase will start;
> if there are errors due to conflicts, git will stop with an error message;
> edit the affected files to resolve the conflicts, then do a git add .; git rebase --continue and go back to 5.;
> once the rebase has completed, use git push --force-with-lease important to force push the changes into your branch
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/betaflight/betaflight/pull/10417#issuecomment-761685205>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ARV2H3E4E7MI47O2QCFLAILS2IDIDANCNFSM4VDIKTUA>.
>

bobbycisneros commented:
Mike,

Corrected the unit test failures.  Please advise if you are going to accept the pull request or if I need to do something else.

Robert

mikeller commented:
@bobbycisneros: I rebased the 29 commits into one for you.
